### PR TITLE
Masking doc and fixes

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Add basic documentation and fix some small issues.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
Do not merge, I believe there is another bug in the masks application logic that I need to fix.

File: docs/user-guide/masking.ipynb [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    "a + b"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Reduction operations\n",
+    "\n",
+    "Operations like `sum` and `mean` can in general not preserve masks.\n",
```
Comment (OwenArnold) [inline docs/user-guide/masking.ipynb]: ambiguous sentence ```Operations like sum and mean can in general not preserve masks```.  ```Operations like sum and mean do not preserve masks```

As a side-note, how could a user find out 
a priori what the masking behaviour of an algorithm is? help for `sum` for example does not describe.

File: docs/user-guide/masking.ipynb [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Operations with masked objects\n",
+    "\n",
+    "### Element-wise binary operations\n",
+    "\n",
+    "The result of operations between data arrays or dataset with masks contains the masks of both inputs.\n",
```
Comment (OwenArnold) [inline docs/user-guide/masking.ipynb]: How about having multiple named masks in the same dataset. For example adding  a second `Dim.X` mask 
```
a.masks['xx'] = sc.Variable([Dim.X], values=[True, False, True])
```
These also seem to be ORed

File: docs/user-guide/masking.ipynb (hunk 1 above)
Comment (SimonHeybrock) [reply docs/user-guide/masking.ipynb]: What is ambiguous?

File: docs/user-guide/masking.ipynb (hunk 2 above)
Comment (SimonHeybrock) [reply docs/user-guide/masking.ipynb]: No they should just be preserved? Will add an example though.

Comment (SimonHeybrock) [reply docs/user-guide/masking.ipynb]: Actually there is an example already, `b` has two masks, `x` and `y`.

File: docs/user-guide/masking.ipynb (hunk 1 above)
Comment (SimonHeybrock) [reply docs/user-guide/masking.ipynb]: Actually, there is an example of this already as well, lower down.

File: docs/user-guide/masking.ipynb (hunk 2 above)
Comment (OwenArnold) [reply docs/user-guide/masking.ipynb]: The second mask for b is in dim.y, not a second mask in the x dimension

Comment (SimonHeybrock) [reply docs/user-guide/masking.ipynb]: That makes no difference?

File: docs/user-guide/masking.ipynb (hunk 1 above)
Comment (OwenArnold) [reply docs/user-guide/masking.ipynb]: "can in general not" doesn't read well, I suggest rewording this as the meaning is not clear.

Comment (SimonHeybrock) [reply docs/user-guide/masking.ipynb]: This is reformulated already in the latest changes.

File: core/include/scipp/core/string.h [hunk 3]
```diff
@@ -83,6 +83,8 @@ template <class T> std::string element_to_string(const T &item) {
   using std::to_string;
   if constexpr (std::is_same_v<T, std::string>)
     return {'"' + item + "\", "};
+  else if constexpr (std::is_same_v<T, bool>)
```
Comment (dtasev) [inline core/include/scipp/core/string.h]: Maybe [this](https://github.com/scipp/scipp/blob/masking-doc/core/include/scipp/core/string.h#L50) overload is not necessary with this?

File: core/variable_operations.cpp [hunk 4]
```diff
@@ -206,12 +206,12 @@ Variable sum(const VariableConstProxy &var, const Dim dim) {
...
 Variable sum(const VariableConstProxy &var, const Dim dim,
              const MasksConstProxy &masks) {
-  if (masks.empty()) {
-    return sum(var, dim);
-  } else {
-    const auto mask_union = masks_merge(masks);
-    return sum(var * ~mask_union, dim);
+  if (!masks.empty()) {
+    const auto mask_union = masks_merge(masks, dim);
+    if (mask_union.dims().contains(dim))
```
Comment (dtasev) [inline core/variable_operations.cpp]: Do we need this check if `masks_merge` filters all masks with unrelated dimensions now?

File: core/variable_operations.cpp [hunk 5]
```diff
@@ -235,13 +235,14 @@ Variable mean(const VariableConstProxy &var, const Dim dim) {
...
               const MasksConstProxy &masks) {
-  if (masks.empty()) {
-    return mean(var, dim);
-  } else {
-    const auto mask_union = masks_merge(masks);
-    const auto masks_sum = sum(mask_union, dim);
-    return mean(var * ~mask_union, dim, masks_sum);
+  if (!masks.empty()) {
+    const auto mask_union = masks_merge(masks, dim);
+    if (mask_union.dims().contains(dim)) {
```
Comment (dtasev) [inline core/variable_operations.cpp]: Same question as above

File: core/include/scipp/core/string.h (hunk 3 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/string.h]: No, it is using that, so it should be necessary?

File: core/variable_operations.cpp (hunk 4 above)
Comment (SimonHeybrock) [reply core/variable_operations.cpp]: Yes, because `mask_union` will be 0-D in that case and we want to skip the multiplication?

File: core/variable_operations.cpp (hunk 5 above)
Comment (SimonHeybrock) [reply core/variable_operations.cpp]: See above, I think it is required.

File: core/include/scipp/core/string.h (hunk 3 above)
Comment (dtasev) [reply core/include/scipp/core/string.h]: My bad. I did not see the name of the function 🤦‍♂

File: core/variable_operations.cpp (hunk 4 above)
Comment (dtasev) [reply core/variable_operations.cpp]: I see. That makes sense

**Comment by SimonHeybrock** [comment]
Owen's comments were regarding documentation improvements, which we can do in the future. Merging.

File: docs/user-guide/masking.ipynb [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    "a + b"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Reduction operations\n",
+    "\n",
+    "Operations like `sum` and `mean` over a particular dimension cannot preserve masks that depend on this dimension.\n",
```
Comment (OwenArnold) [inline docs/user-guide/masking.ipynb]: 👍

File: docs/user-guide/masking.ipynb [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+    "If this is the case, the mask is applied during the operation and is not present in the output:"
```
Comment (OwenArnold) [inline docs/user-guide/masking.ipynb]: I think the description and examples here are good and clarify an existing assumption I had with the declaration of the masks - that they might broadcast over the other dimensions.
